### DISCUSSION TITLE
[@types/async] - forEachSeries and other methods didn't typed correctly for async iterators

### DISCUSSION THREAD

**Discussion by npdev453** [DefinitelyTyped/DefinitelyTyped#64814] [Issues with a @types package]
Seems like forEachSeries and other methods didn't typed correctly when iterators is async:

```ts
async.forEachSeries(array, async (item) => {
     await testPromise()
...
})
```
Second arg will have there wrong signature `(err, cb)=>void` instead of `(err)=>Promise`

That will cause an eslint error `@typescript-eslint/no-misused-promises` and also have incorrect signature in type-tip

I think there are should be something like additional type like `AsyncResultIteratorPromise`  for iterator, not only `AsyncIterator`

____
PS:

Also callback is really optional there, when it missed, async-lib will `await`ed of iterrator and `cb` will be undefined. 

I mean that this code will cause an error in runtime, but not on type-checking stage:
```ts
async.forEachSeries(array, async (item, next) => {
     await testPromise()
     next()
})
```

Runtime error:
`Uncaught TypeError: next is not a function`

[a comment by an automated account is not shown]

**Comment by npdev453** [comment]
seems like reated to https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/59165
but not fixed enought

**Comment by npdev453** [comment]
@joaop21 , maybe you can be interested?

**Reply by joaop21** [reply]
Yes, I am, indeed. I tried some things on my side, but my typescript knowledge is not the best for the job. I can't make the second parameter (the callback) give a type error when the return type is `Promise<void>`.
Tbh, most of these package functions have the same problem, but they're not well-typed.

**Reply by npdev453** [reply]
>  Tbh, most of these package functions have the same problem, but they're not well-typed.

I'm trying also too, but got failed. Seems like there blocked issue:  (that does't allow to make next arg optional if promise case)
